signalfd.c

```diff
@@ -0,0 +1,170 @@
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "event2/event-config.h"
+
+#include <unistd.h>
+#include <sys/signalfd.h>
+
+#include "event2/event.h"
+#include "event-internal.h"
+#include "evmap-internal.h"
+#include "evsignal-internal.h"
+#include "evthread-internal.h"
+
+static int sigfd_add(struct event_base *, evutil_socket_t, short, short, void *);
+static int sigfd_del(struct event_base *, evutil_socket_t, short, short, void *);
+
+static const struct eventop sigfdops = {
+	"signalfd_signal",
+	NULL,
+	sigfd_add,
+	sigfd_del,
+	NULL,
+	NULL,
+	0, 0, 0
+};
+
+static void
+sigfd_cb(evutil_socket_t fd, short what, void *arg)
+{
+	struct signalfd_siginfo fdsi;
+	struct event_base *base = arg;
+	ssize_t ret = read(fd, &fdsi, sizeof(fdsi));
+
+	EVUTIL_ASSERT(ret == sizeof(fdsi));
+	EVUTIL_ASSERT(fdsi.ssi_signo > 0 && fdsi.ssi_signo < NSIG);
+	EVUTIL_ASSERT(base->sig.ev_sigevent[fdsi.ssi_signo] != NULL);
+
+	EVBASE_ACQUIRE_LOCK(base, th_base_lock);
+	evmap_signal_active_(base, fdsi.ssi_signo, 1);
+	EVBASE_RELEASE_LOCK(base, th_base_lock);
+}
+
+static void
+sigfd_free_sigevent(struct event_base *base, int signo)
+{
+	int ret;
+	struct event* sigev = base->sig.ev_sigevent[signo];
+
+	EVUTIL_ASSERT(sigev != NULL);
+	/* Not sure about blocking here. */
+	event_del_nolock_(sigev, EVENT_DEL_AUTOBLOCK);
+	ret = close(sigev->ev_fd);
+	EVUTIL_ASSERT(ret == 0);
+	mm_free(sigev);
+	base->sig.ev_sigevent[signo] = NULL;
+}
+
+static int
+sigfd_add(struct event_base *base, int signo, short old, short events, void *p)
+{
+	int sigfd;
+	sigset_t mask;
+	struct event* sigev;
+
+	/* EV_SIGNAL event passed from evmap_signal_add() when setting
+           up and from evmap_signal_reinit_iter_fn() during reinit. */
+	EVUTIL_ASSERT(p != NULL);
+
+	EVUTIL_ASSERT(signo > 0 && signo < NSIG);
+	sigev = base->sig.ev_sigevent[signo];
+
+	if (sigev != NULL) {
+		if (old) {
+			/* We're performing reinit after fork(). */
+			sigfd_free_sigevent(base, signo);
+		} else {
+			/* We have an active signal fd
+			   for this signal already. */
+			return 0;
+		}
+	}
+
+	sigemptyset(&mask);
+	sigaddset(&mask, signo);
+	if (sigprocmask(SIG_BLOCK, &mask, NULL)) {
+		event_warn("sigprocmask() failed");
+		return -1;
+	}
+
+	sigfd = signalfd(-1, &mask, SFD_NONBLOCK);
+	if (sigfd < 0) {
+		event_warn("signalfd() failed");
+		goto unblock;
+	}
+
+	/* EV_READ event used to wakeup corresponding EV_SIGNAL ones. */
+	sigev = event_new(base, sigfd, EV_READ | EV_PERSIST, sigfd_cb, base);
```

Comment:
There is also some trickery with polling fd after fork, from `signalfd(2)`:

```
   epoll(7) semantics
       If a process adds (via epoll_ctl(2)) a signalfd file descriptor to an epoll(7) instance, then epoll_wait(2) returns events only for signals sent to that process.  In par‐
       ticular, if the process then uses fork(2) to create a child process, then the child will be able to read(2) signals that are sent to it using the signalfd  file  descrip‐
       tor, but epoll_wait(2) will not indicate that the signalfd file descriptor is ready.  In this scenario, a possible workaround is that after the fork(2), the child process
       can close the signalfd file descriptor that it inherited from the parent process and then create another signalfd file descriptor and add it to the epoll  instance.   Al‐
       ternatively,  the  parent  and  the  child  could  delay creating their (separate) signalfd file descriptors and adding them to the epoll instance until after the call to
       fork(2).
```

So if I understand correctly then the child cannot poll the descriptor, only read from it.
And we are safe here because of `evmap_signal_reinit_iter_fn`, let's put this in the comment in code.

Reply:
OK